nimble/host/src/ble_sm.c

```diff
@@ -2910,4 +2910,109 @@ ble_sm_create_chan(uint16_t conn_handle)
     return chan;
 }
 
+#if MYNEWT_VAL(BLE_SM_SC)
+int
+ble_sm_csis_decrypt_sirk(const uint8_t *ltk, const uint8_t *enc_sirk, uint8_t *out)
+{
+    int rc;
+
+    /* Decrypt SIRK with sdf(K, EncSIRK)  */
+    rc = ble_sm_alg_csis_sdf(ltk, enc_sirk, out);
+
+    return rc;
+}
+
+int
+ble_sm_csis_resolve_rsi(ble_addr_t peer_addr, uint8_t *rsi,
+                        uint8_t *sirk, bool is_encrypted)
+{
+    struct ble_store_key_sec key_sec;
+    struct ble_store_value_sec value_sec;
+
+    uint8_t plaintext_sirk[16] = {0};
+    uint8_t local_hash[3] = {0};
+    uint8_t prand[3] = {0};
+    uint8_t hash[3] = {0};
+    int rc;
+
+    memset(&key_sec, 0, sizeof(key_sec));
+    key_sec.peer_addr.type = peer_addr.type;
+    memcpy(key_sec.peer_addr.val, peer_addr.val, 6);
+
+    rc = ble_store_read_peer_sec(&key_sec, &value_sec);
+    if (rc != 0) {
+        return rc;
+    } else if (!value_sec.ltk_present) {
+        return BLE_HS_ENOENT;
+    }
+
+    memcpy(hash, rsi, 3);
+    memcpy(prand, rsi + 3, 3);
+
+    if (is_encrypted) {
+        rc = ble_sm_csis_decrypt_sirk(value_sec.ltk, sirk, plaintext_sirk);
+        if (rc != 0) {
+            return rc;
+        }
+    } else {
+        memcpy(plaintext_sirk, sirk, 16);
+    }
+
+    rc = ble_sm_alg_csis_sih(plaintext_sirk, prand, local_hash);
+    if (rc != 0) {
+        return rc;
+    }
+
+    if (memcmp(local_hash, hash, 3) == 0) {
```

Comment:
better write this as:
```c
if (memcmp(local_hash, hash, 3)) {
    return BLE_HS_EAUTHEN;
}

return 0;
```

this way return value in case of success is at the very end of function as expected